_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+This post is to clear things up.
+<!--more-->
+
+This post will not cover anything in the [Julia Performance Tips](https://docs.julialang.org/en/v1/manual/performance-tips/).
+
+Each section in this will describe a different antipattern.
+It will discuss:
+ - Examples, including highlighting this issues
+ - What to do instead
+ - Reasons people do this
+
+## NotImplemented Exceptions
+
+This shows up in packages defining APIs that others should implement.
+Where one wants to declare and document functions that implementors of the API should overload.
+In this antipattern, one declares an abstract type, and a function that takes that abstract type as an argument.
+Usually, as the the first abstract following something like fashion from Python etc, as taking the object as the first argument.
+This function would throw an error saying that the function was not implemented for this type.
+
+The logic being if someone only implements half the API, the user would get this "Helpful" error message.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+abstract type AbstractModel end
+
+"""
+    probability_estimate(model, observation::AbstractArray)::Real
+
+For a given `model`, returns the likelihood of the `observation` occurring.
+"""
+function probability_estimate(model::AbstractModel, observation::AbstractArray)
+    error("`probability_estimate` has not been implemented for $(typeof(model))")
+end
+
+## Now a type implementing this API:
+"""
+    GuessingModel <: AbstractModel
+
+A model that just guesses. Not even educated guesses. Just random guessing.
+"""
+struct GuessingModel <: AbstractModel
+end
+
+probability_estimate(guesser::GuessingModel, observation::AbstractMatrix) = rand()
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+probability_estimate (generic function with 2 methods)
+```
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+probability_estimate(GuessingModel(), [1,2,3])
+```
+</div>
+
+**Output:**
+
+    `probability_estimate` has not been implemented for GuessingModel
+
+**What happened?**  
+It sure looks like `GuessingModel` has implemented a `probability_estimate` method.
+The error does not at all help us see what is wrong.
+Astute eyed readers will see what is wrong:
+`GuessingModel()` was incorrectly implemented to only work for `AbstractMatrix`, but it was called with a `Vector`,
+ so it fell back to the generic method for `AbstractModel`.
+But that error message what not informative, and if we hit that deep inside a program we would have no idea what is going on, because it doesn't print the types of all the arguments.
+
+### What to do instead?
+Just don't implement the thing you don't want to implement.
+A `MethodError` indicates this quite well and as shown gives a more informative error message than you will write.
+
+A often missed feature is that you can declare a function without providing any methods.
```

Comment:
I would use overlooked here,  instead of missed, because missed could also mean, that someone wants this feature and this confused me just now. 
Also it should be An often